Compile your Node.js application into a .exe file (#2615)
section-engineering-education/engineering-education: content/articles/compile-your-node.js-application-into-a-.exe-file/index.md

> During program development, one undergoes several processes of installation, coding, and running an application.
> These processes drastically change during deployment or execution where the users are just interested in the running of the application and not the underlying codes.
> This arises the need to creation of a simple executable file that can run on the current operating system without further package installation process.
> 
> In this tutorial, we will learn how to create executable files for our Node.js application. We shall compile our JavaScript files into executable files.

Reviewer: ```suggestion
During the development phase, the developer performs several processes of installation and coding, to build and run an application. But, the end-users are just interested in running the application and not the underlying code and processes.

This causes the need for a simple executable file that can run on an operating system without requiring the user to perform other steps in order to run the application.

In this tutorial, we'll learn how to create executable files for our Node.js application. We'll compile our JavaScript files into an executable file.
```

- [A simple web app](#A-simple-web-app "A simple web app")
- [nexe](#nexe "nexe")
    [nexe installation](#nexe-installation "nexe installation")
    [Compile one javascript file into an exercutable using nexe](#Compile-one-javascript-file-into-an-exercutable-using-nexe "Compile one javascript file into an exercutable using nexe")
    [Compile your project into an exercutable using nexe](#Compile-your-project-into-an-exercutable-using-nexe "Compile your project into an exercutable using nexe")
- [pkg](#pkg "pkg")
    [pkg installation](#pkg-installation "pkg installation")
    [Compile one javascript file into an exercutable using pkg](#Compile-one-javascript-file-into-an-exercutable-using-pkg "Compile one javascript file into an exercutable using pkg")
    [Compile your project into an exercutable using pkg](#Compile-your-project-into-an-exercutable-using-pkg "Compile your project into an exercutable using pkg")
    [File or directory not included Error](#File-or-directory-not-included-Error)
- [Further practice](#Further-practice "Further practice")
- [References](#References "References")

### Key takeaways

At the end of this tutorial, you would have learned the following:

> - What are `.exe` files and what are their importance.
> - Some packages we can use to compile your Node.js application into `.exe` files.
> - Creating a simple Node.js application.
> - Compiling Node.js files into an executable file for different operating systems.
> - Running the executable file in the current operating system.

Reviewer: ```suggestion
- What are `.exe` files and their importance.
- Packages that we can use to compile your Node.js application into `.exe` files.
- Creating a simple Node.js application.
- Compiling Node.js files into an executable file for various operating systems.
- Running the executable file in an operating system.
```

### Pre-requisites

For one to easily follow up with this tutorial, you require a minimum of the following requirements:

> - Basic web development skills.
> - Some javascript coding skills.
> - Node.js installed in your machine. If not, download it [here](https://nodejs.org/en/download/ "Node.js download").
> - An IDE. I highly recommend Visual Studio Code which is free and highly customizable. You can download it [here](https://code.visualstudio.com/download "Visual Studio Code download").
> - A stable internet connection.

Reviewer: ```suggestion
- Basic web development skills.
- Some javascript coding skills.
- Node.js installed in your machine. If not, download it from their [website](https://nodejs.org/en/download/).
- An IDE. I highly recommend Visual Studio Code which is free and highly customizable. You can download it from [here](https://code.visualstudio.com/download).
- A stable internet connection.
```

If you are all set on the above requirements, let us look at some packages in which we can use to accomplish our goal.

### What are Executable (.exe) files

Executable files (.exe files): These are files containing a set of encoded instructions which are executed sequentially once the user clicks on the file or runs it.

> `.exe` is a Windows extension referring to an executable file.
> Equivalent to the `.exe files` in linux or Unix systems are the files ending with the `.bin`, `.elf` or `[none]` extensions.
> In macOS, they lack extensions.

Reviewer: ```suggestion
`.exe` is a Windows extension referring to an executable file. In Linux or Unix-based operating systems, files ending with the `.bin`, `.elf` or `[none]` extensions are executables. In macOS, they lack extensions.
```

### Pros of executable files

Let us first see why prefer executable files in the first place. Some advantages noted of executable files include:

> - **Quick no-code execution:** This is because it only requires one to run the file and that's it. No codes are needed to be run on the terminal for it to be deployed.
> - **Prevent unwanted code change:** It prevents unwanted errors generated due to unplanned and accidental modification of source code.
> - **Supports distribution:** You can run it on any machine. One doesn't need to install any packages or dependencies, just running it. This allows one to run multiple instances of the code on any machine without further requirements.

Reviewer: ```suggestion
- **Quick no-code execution:** It only requires one click to run the file and that's it. Commands are not required to be run on the terminal to execute the program.

- **Prevent unwanted code change:** It prevents unwanted errors generated due to unplanned and accidental modification of source code.

- **Supports distribution:** You can run it on any machine. The user doesn't have to install any packages or dependencies. This allows one to run multiple instances of the code on any machine without further requirements.
```

### Packages used

> Some common two packages used to compile javascript files into executables include:

Reviewer: ```suggestion
Two of the most commonly used packages used to compile javascript files into executables are:
```

> - [nexe](https://www.npmjs.com/package/nexe "npmjs.com nexe/nexe documentation"): It is a simple command-line utility that compiles your Node.js application into a single executable file. By default, it converts it into a windows format. You can quickly jump over to its section by clicking [here](#nexe "nexe").

Reviewer: ```suggestion
- [nexe](https://www.npmjs.com/package/nexe "npmjs.com nexe/nexe documentation"): It is a simple command-line utility that compiles your Node.js application into a single executable file. By default, it converts it into a Windows executable. You can quickly jump over to its section by clicking [here](#nexe "nexe").
```

> - [pkg](https://www.npmjs.com/package/pkg): It is a node package that can convert your node app into several executable files for different Operating systems (all at once) or of an individual  O/S. You can quickly jump over to its section by clicking [here](#pkg "pkg").

Reviewer: ```suggestion
- [pkg](https://www.npmjs.com/package/pkg): It is a node package that can convert your node app into several executable files for various operating systems (all at once) or of an individual operating system. You can quickly jump over to its section by clicking [here](#pkg "pkg").
```

We shall now get a simple node.js application that is running and use it as our application.

### A simple web app

> This web app shall be able to run on localhost at port 5000 and shall display images found in the images folder. Create a route directory called "executable" and create files and folders as shown in the folder structure below.

Reviewer: ```suggestion
This web app will run on localhost at port 5000 and will display images found in the images folder. Create a directory called "executable" and create files and folders as shown in the folder structure below.
```

Initialize the project with:

```bash
npm init -y
```

Install the packages needed for the project using:

```bash
npm i express
```

#### Folder Structure

The app will look as shown below:

```bash
├──  node_modules (folder)
├──  views (folder)
│   └── images (folder)
│       └── tv1.jpg (file)
│   └── index.html (file)
│   └── sakura.css (file)
├──  index.js (file)
└──  package.json (file)
```

#### views

##### images

> Inside the image folder, download a free 760 × 380 tv image from the web and name it "tv1.jpg".

Author: The resolution is not important.
Ok will do

> Inside the image folder, download a free 760 × 380 tv image from the web and name it "tv1.jpg".

Author: Done!

##### index.html

The code is as follows:

```html
<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Exercutable Files</title>
    <link rel="stylesheet" href="sakura.css">
</head>

<body>
    <!--Display text and a tv image-->
    <h1>Hello, world!</h1>
    <br>
    <p>Its you again, Hello there!</p>
    <div class="photo">
        <h2>TV 1</h2>
        <img src="./images/tv1.jpg" alt="tv1" class="img">
        <br>
        <hr>
        <h2>TV 2</h2>
        <img src="./images/tv1.jpg" alt="tv2" class="img">
    </div>
</body>
</body>

</html>
```

> ##### sakura.css

Author: Ok, Done!

app.get('/', function(req, res) {
    res.sendFile(getDir() + '/views/index.html');
});

// Using a function to set default app path
function getDir() {
    if (process.pkg) {
        return path.resolve(process.execPath + "/..");
    } else {
        return path.join(require.main ? require.main.path : process.cwd());
    }
}
```

> **Note:** Use the `getDir()` function or `path.join` instead of `__dirname` when using express or fastify routings to avoid errors.

#### package.json

The code automatically generated will look as shown below:

```json
{
    "name": "executable",
    "version": "1.0.0",
    "description": "Simple express app",
    "main": "index.js",
    "scripts": {
        "start": "node index.js",
        "test": "echo \"Error: no test specified\" && exit 1"
    },
    "keywords": [],
    "author": "",
    "license": "ISC",
    "dependencies": {
        "express": "^4.17.1"
    }
}
```

Open the inbuilt Visual Studio Code Terminal using ``Ctrl + ` `` in PC or ``Control + Shift + ` `` in Mac.

Run the application using the following command on the terminal to get "Server at 5000" output using:

```bash
node index.js
```

> Open a browser and in the URL section access the app using `localhost:5000`. This shows it is functioning.

Reviewer: ```suggestion
Open a browser and open the app by visiting `localhost:5000`.
```

Once done, let's see how each module operates.

### nexe

![nexe](/engineering-education/compile-your-node.js-application-into-a-.exe-file/nexe.png "nexe")

Now let us compile our project into an executable file with all needed packages and resources. Apart from the "node_modules", the other resource required that is also found in the root directory is the "views" folder. The node packages will be automatically added, so we will only specify other resources needed.

#### nexe installation

Run in the Terminal the following to install `nexe` globally:

```bash
npm i nexe -g
```

If you are using Ubuntu, run:

```bash
sudo npm i nexe -g
```

> **Note:** Unlike in the pkg module, we don't install the package locally but rather globally. If you install it locally, you will face numerous errors. An example of that error is `Command 'nexe' not found`. You can see how to resolve it [here](https://stackoverflow.com/questions/67981535/nexe-module-command-nexe-not-found-error/67981536#67981536 "Stack Overflow command 'nexe' not found Error").

> Once done, run the following to verify its availability:

Reviewer: ```suggestion
Once done, run the following to verify if its installed:
```

```bash
nexe -h
```

#### Compile one javascript file into an executable using nexe

Now, we shall run the following to just convert the application entry point into an executable file:

```bash
nexe index.js
```

> This tells `nexe` that during the build process, it will only look and compile the "index.js" file into an executable file.

Reviewer: ```suggestion
This tells `nexe` that during the build process, it should only look and compile the "index.js" file into an executable file.
```

Once done run the following command on the terminal to start the build process:

```bash
nexe --build
```

> It will create a new executable file with the same name as our root directory's name "executable". By default, it will use the system O/S and architecture type as the targets for the system so that you can quickly run it on your system.
> Run it using the command below on Ubuntu:

Reviewer: ```suggestion
I'll create a new executable file with the same name as our root directory's name "executable". By default, it'll use the system's operating system and architecture type as the target for the application so that you can quickly run it on your system.

Run the generated executable using the command below on Ubuntu:
```

```bash
./executable
```

It automatically logs out "Server at 5000" meaning that it is running. Access the application in the preferred browser at `localhost:5000`.
You can close it using `Ctrl + C`.

> To prove that it is only the entry point is compiled and not the resources found in the root directory such as the ones in the "views" folder, just copy and paste the executable file in a separate folder and run it. It pops up errors since it is unable to access the resources.

Reviewer: ```suggestion
To prove that only the entry point is compiled and not the resources found in the root directory such as the ones in the "views" folder, just copy and paste the executable file in a separate folder and run it. It pops up errors since the application will be unable to access the resources.
```

> #### Compile your project into an executable using nexe

Reviewer: ```suggestion
#### Compile the entire project into an executable using nexe
```

We can do this in two ways.

> - In the first, we can add the resources in the command line using the `--resources` or `-r` as seen below:

Reviewer: ```suggestion
- First, we can add the resources in the command line using the `--resources` or `-r` flag as seen below:
```

```bash
nexe -r "views/**/*"
```

We can also add additional commands such as the output executable file path using `--output` or `-o`, target using `--target` or `-t`, name of the executable file using `--name` or `-n`, build from source using `--build` or `-b` e.t.c.

> 'target' in `nexe` is used to specify the platform (windows, linux, or macos), the arch or architecture type (x86, x64), and the node version (12, 14, 16). You can see more [here](https://github.com/nexe/nexe "nexe Readme").

> - In the second, we can define the resources and others inside the 'package.json' file. I recommend this since your configurations would be saved for later re-run. Since we already declared that our app "main" is "index.js", we shall add only a build script.

Reviewer: ```suggestion
2. We can define the resources inside the 'package.json' file. I recommend this since your configurations would be saved for later re-run. Since we already declared that our app "main" is "index.js", we shall add only a build script.
```

Head over to the "package.json" file and under the scripts tag, add a build script as follows:

```json
{
    "scripts": {
        "start": "node index.js",
        "build": "nexe -r views/**/*",
        "test": "echo \"Error: no test specified\" && exit 1"
    }
}
```

> This shall specify our resources which are locally found during the build time.

Reviewer: ```suggestion
This will specify our resources which are locally found during the build time.
```

Now run the command below to tell `nexe` to use the specifications in the build tag:

```bash
npm run build
```

> then to run the build process we shall use:

Reviewer: ```suggestion
To start the build process, use the following command:
```

```bash
npm --build
```

> You can run the file created. You can also copy it to another different directory without any resources and try running it to verify that the resources were compiled into the executable file.

Reviewer: ```suggestion
You can run the file that's created. You can also copy it to another directory without any resources and try running it to verify that the resources were compiled into the executable file.
```

> In case of further errors, you can post them in the [nexe issues tab](https://github.com/nexe/nexe/issues "nexe issues tab") for help.

Reviewer: ```suggestion
In case of further errors, you can post them in the [nexe issues tab](https://github.com/nexe/nexe/issues) for help.
```

****

### pkg

![pkg](/engineering-education/compile-your-node.js-application-into-a-.exe-file/nexe.png "pkg")

> Let us use the project previously created and see the outcome.

Reviewer: ```suggestion
Let's use the project that we previously created and see the outcome.
```

> You can either delete the contents of the root directory (executable) previously created or create a new one with a different name and repeat the steps we did above to create a new project. This reduces vulnerability errors due to packages clashing with each other.

#### pkg installation

You can start by installing the package using the command below for local installation:

```bash
npm i pkg
```

or for global installation use:

```bash
npm i pkg -g
```

Wait for a few seconds for it to complete the process. On completion, just check and close any instance in which the code is running in the background.

#### Compile one javascript file into an executable using pkg

> Run the following command in which we shall call the pkg module and the entry point to our application. In our situation it is `index.js`:

Reviewer: ```suggestion
Run the following command in which we'll call the pkg module and the entry point to our application. In our case, its `index.js`:
```

```bash
pkg index.js
```

> This will only compile our app entry point into three different executable files since no targets were specified.

Reviewer: ```suggestion
This will compile our app's entry point into three different executable files since no targets were specified.
```

Just wait for some time for the process to complete. Head over to the main directory. You will find three newly generated files. These are the:

- for Windows (index-win.exe)
- for Linux (index-linux)
- for mac (index-macos)

> That is because we did not specify one O/S.
> Run the file suitable for your O/S.
> If you are using Ubuntu, run `./index-linux` on the terminal.
> It will launch the app and display a console that will act as an interactive interface between you and the app. Access the application in the preferred browser at `localhost:5000`. As seen, it works well.

Reviewer: ```suggestion
That is because we did not specify an operating system. Run the file that's suitable for your operating system. If you are using Ubuntu, run `./index-linux` on the terminal. It will launch the app and display a console that will act as an interactive interface between you and the app. Access the application in the preferred browser at `localhost:5000`. As seen, it works well.
```

> **NOTE:** As long as the executable file is running, the application is running.

Close it using `Ctrl + C`.

Now let us dive a little deeper and specify the operating system and the node version.

> In this, we use the `--targets` command.

Reviewer: ```suggestion
In this, we use the `--targets` flag to set the target operating system.
```

Delete the executable files generated in the previous process.

Run the following in the Terminal:

```bash
pkg index.js --targets node12-win-x64
```

This specifies that the project should be compiled into an executable file that runs on Node version 12 and Windows O/S of a 64-bit architecture.

Some of the supported node ranges, platforms, and architectures are as shown in the table below:

| Node Versions | Platforms  | Archs  |
|---|---|---|
| node8  | alpine  | x64  |
| node10  | linux  | arm64  |
| node12  | linuxstatic  | (armv6)  |
| node14  | win  | (armv7)  |
| node16  | macos  |   |
| latest  | (freebsd)  |   |

#### Compile your project into an executable using pkg

> Here, we shall have to tell `pkg` our resources folder. We can do this inside the `package.json` file using scripts and assets configurations.

Reviewer: ```suggestion
Here, we should tell `pkg` our resources folder. We can do this inside the `package.json` file using scripts and assets configurations.
```

Let us head over to our `package.json` file and add a "pkg" as shown below:

```json
{
    "name": "executable",
    "bin": "index.js",
    "version": "1.0.0",
    "description": "Simple express app",
    "main": "index.js",
    "scripts": {
        "start": "node index.js",
        "test": "echo \"Error: no test specified\" && exit 1"
    },
    "pkg": {
        "assets": [
            "views/**/*"
        ],
        "output": "dist"
    },
    "keywords": [],
    "author": "",
    "license": "ISC",
    "dependencies": {
        "express": "^4.17.1",
        "pkg": "^5.2.1"
    }
}
```

> You can notice we have also added "bin". This tells `pkg` where the app entry point is. Inside the "pkg" script, we have added "assets" to show which file contains our resources. You can add more by just separating them with a comma.  We have further specified where the outputs will be stored using "outputPath".

Reviewer: ```suggestion
Notice that we have also added "bin". This tells `pkg` where the app entry point is. Inside the "pkg" script, we have added "assets" to show which file contains our resources. You can add more by just separating them with a comma.  We have further specified where the outputs will be stored using "outputPath".
```

Run `pkg` using the command below to use our configuration in `package.json`:

```bash
pkg .
```

or

```bash
pkg package.json
```

In the case of build files or the specific targets, you can also specify them to pkg to obtain an output of the format below:

```json
{
    "pkg": {
              "scripts": "build/**/*.js",
              "assets": "views/**/*",
              "targets": [ "node14-linux-arm64" ],
              "outputPath": "dist"
            }
}
```

#### File or directory not included Error

> Incase you might have encountered "was not included" Error in pkg, then this is how to get around it.

Reviewer: ```suggestion
In case you have encountered the "was not included" Error in pkg, then this is how to get around it.
```

> In pkg module, we can not use direct concatenation of relative paths, for instance `res.sendFile(__dirname + '/views/index.html');`. This is because it is not only a **bad programming practice** but also because it will retun an error as shown below:

Reviewer: ```suggestion
In pkg module, we shall not use a direct concatenation of relative paths, for instance `res.sendFile(__dirname + '/views/index.html');`. This is because it is not only a **bad programming practice** but also because it will return an error as shown below:
```

Instead of using `__dirname`, replace it with either `path.join` or `getDir()` in which you have required *path* at the beginning of the file using `const path = require('path');`. `process.cwd()` is always used for files which will not be available during build time but would be required for execution.
In case of more issues you can read further [here](https://stackoverflow.com/questions/67966111/node-js-vercel-pkg-express-return-0-error-and-fastify-errors-error-file-or-f/67966112#67966112 "Stack-Overflow vercel/pkg express and fastify errors"). You can raise up newer issues [here](https://github.com/vercel/pkg/issues "vercel/pkg issues tab").

### Further practice

- Check out additional functions at [nexe/nexe](https://github.com/nexe/nexe "Github.com nexe/nexe") and [vercel/pkg](https://github.com/vercel/pkg "Github.com vercel/pkg") and try them out.
- You can try running the other two files in their respective virtual machines and see the outcome.
Just install the other two operating systems and configure them. Run the files and test the outcome on web browsers of choice.

> > As stated before you can copy it to another machine and the other person won't need to do any package installation process. It will just run them.

Reviewer: ```suggestion
> As stated before you can copy it to another machine and the other person won't need to do any package installation process.
```

    - [File or directory not included Error](#File-or-directory-not-included-Error)
- [Further practice](#Further-practice "Further practice")
- [References](#References "References")

### Key takeaways

At the end of this tutorial, you would have learned the following:

- What are `.exe` files and their importance.
- Packages that we can use to compile your Node.js application into `.exe` files.
- Creating a simple Node.js application.
- Compiling Node.js files into an executable file for various operating systems.
- Running the executable file in an operating system.

### Pre-requisites

For one to quickly follow up with this tutorial, you require a minimum of the following requirements:

- Basic web development skills.
- Some javascript coding skills.
- Node.js installed in your machine. If not, download it from their [website](https://nodejs.org/en/download/).
- An IDE. I highly recommend Visual Studio Code which is free and highly customizable. You can download it from [here](https://code.visualstudio.com/download).
- A stable internet connection.

If you are all set on the above requirements, let's look at some packages to accomplish our goal.

> ### What are Executable (.exe) files

Reviewer 2: You have to specify if we are only working with Windows-based executable files.

Author: Ok.
The main purpose of this article was intended for windows but the packages support a variety of other O/S so I thought it would be better to mention it as a bonus! So should I change it?

Author: Ok let me get into it

Author: Please check to see if it is as expected

- **Prevent unwanted code change:** It prevents unwanted errors generated due to unplanned and accidental modification of source code.

- **Supports distribution:** Being hardware agnostic, the user doesn't have to install any packages or dependencies. This allows one to run multiple instances of the code on any machine without further requirements.

### Packages used

Two of the most commonly used packages used to compile javascript files into executables are:

- [nexe](https://www.npmjs.com/package/nexe "npmjs.com nexe/nexe documentation"): It is a simple command-line utility that compiles your Node.js application into a single executable file. By default, it converts it into a Windows executable. You can quickly jump over to its section by clicking [here](#nexe "nexe").
- [pkg](https://www.npmjs.com/package/pkg): It is a node package that can convert your node app into several executable files for various operating systems (all at once) or of an individual operating system. You can quickly jump over to its section by clicking [here](#pkg "pkg").

> We shall now get a simple node.js application that is running and use it as our application.

Reviewer 2: Node.js. Use the proper syntax when referring to Node.js throughout the article.

Author: Ok.

Author: Done!
